### PR TITLE
Configuration restful API

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
# Description

## Issue reference

Please reference the issue this PR will close: #_[issue number]_
https://github.com/dapr/dapr/issues/4528

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [x] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [x] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
@LaurenceLiZhixin please fix the linter issues: https://github.com/dapr/dapr/runs/6218476858?check_suite_focus=true.

**Comment by yaron2** [comment]
@LaurenceLiZhixin there's a failure: `undeclared name: `jsoniter` (typecheck)`.

File: pkg/http/api.go [hunk 1]
```diff
@@ -677,6 +710,213 @@ func (a *api) onGetState(reqCtx *fasthttp.RequestCtx) {
...
+
+		policy := h.res.ComponentInboundPolicy(ctx, h.storeName)
+		err := policy(func(ctx context.Context) (err error) {
+			resp, err := h.appChannel.InvokeMethod(ctx, req)
+			if err != nil {
+				return err
+			}
+
+			if resp != nil && resp.Status().Code != nethttp.StatusOK {
+				return errors.Errorf("Error sending binding event to application, status %d", resp.Status().Code)
```
Comment (yaron2) [inline pkg/http/api.go:758]: Change `binding event` to `configuration item` please.

File: pkg/http/api.go [hunk 2]
```diff
@@ -677,6 +710,213 @@ func (a *api) onGetState(reqCtx *fasthttp.RequestCtx) {
...
+			}
+			return nil
+		})
+		if err != nil {
+			log.Error(errors.Wrap(err, "error push subscribe event to app"))
```
Comment (yaron2) [inline pkg/http/api.go:763]: Change to `error sending configuration item to the app`.

**Review by yaron2 — CHANGES_REQUESTED** [review]

**Review by yaron2 — APPROVED** [review]

**Comment by yaron2** [comment]
@LaurenceLiZhixin can you please open a PR that updates the API reference with the new endpoints? https://docs.dapr.io/reference/api/

**Comment by msfussell** [comment]
@LaurenceLiZhixin - Can you also open an issue and a  PR for docs updates as well as the reference update.

These two article need to be updated, especially the code examples showing HTTP.
https://docs.dapr.io/developing-applications/building-blocks/configuration/configuration-api-overview/
https://docs.dapr.io/developing-applications/building-blocks/configuration/howto-manage-configuration/

**Comment by LaurenceLiZhixin** [comment]
OK, I would do it as soon as I can.

**Comment by msfussell** [comment]
Looking at the docs PR for this I would have expected this to have a key value in the URL and not in the query parameters,
to be consistent with the state management GET API. Like this?

GET  http://localhost:<daprPort>/v1.0/configuration/<storename>/<key>
example:
GET http://localhost:3500/v1.0/configuration/myConfigurationStore/myConfigKey

rather than this
GET http://localhost:3500/v1.0/configuration/myConfigurationStore?key=myConfigKey&metadata.contentType=application/json

**Comment by LaurenceLiZhixin** [comment]
In the configuration get request, a key is not required. You can pass one key in a request, or multiple keys, or no keys to get the entire configuration. I think query param is more appropriate. @msfussell

**Comment by yaron2** [comment]
> 

That's correct. Thanks for the explanation.

**Pull request merged**